### DISCUSSION TITLE
Custom transport with changing networks

### DISCUSSION THREAD

**Discussion by 0xsambugs** [wevm/viem#385] [Question]
Hey everyone!

I have a question about how the custom transport would work with a fallback transport in regards with network changes.

Let's say I have two transports:
- A `custom` transport connected to Metamask that is currently connected to Optimism
- A `http` transport connected via Alchemy, also to Optimism

Then I set up a `fallback` transport with both of them, so that if one fails, the other can be used. If I execute `getChainId`, then I would be getting `10` (Optimism chain id) as a return value

Now, if I switch the network on my Metamask to let's say Arbitrum, and I execute `getChainId`, I could be getting either Optimism's or Arbitrum's chain id in return, right?

Do you have any suggestions about something we could do to mitigate this situation? We thought that maybe we could build new transports when the network changed, but our server started crashing with out of memory issues. I **think** that since viem has some polling and timeouts, the previous/old transport is not deleted by the garbage collector

Anyways, love the product and thank you for your time 🙏

**Comment by jxom** [comment]
Hmm yeah, we might need to add an `eth_chainId` check to the `custom` transport (and throw if on wrong chain so it falls back) so it can work nicely with the fallback transport. Can take a look!

**Reply by jxom** [reply]
> We thought that maybe we could build new transports when the network changed, but our server started crashing with out of memory issues

Can you see if this still happens when you set `{ rank: false }` as an option to the `fallback` transport?

**Reply by 0xsambugs** [reply]
> Hmm yeah, we might need to add an `eth_chainId` check to the `custom` transport (and throw if on wrong chain so it falls back) so it can work nicely with the fallback transport. Can take a look!

That would be amazing! Thank you @jxom 🙏 

> Can you see if this still happens when you set `{ rank: false }` as an option to the `fallback` transport?

Will test it out and get back to you! 🫡

**Reply by 0xsambugs** [reply]
Hey @jxom , I have some metrics. The test was basically start our API, and ping (every 5 seconds) one of our endpoints for each network we support: polygon, arbitrum, optimism, ethereum. This endpoint performs a multicall, and that's pretty much it

This is with `{ rank: false }`. Memory and active handles remain constant:
<img width="852" alt="Screenshot 2023-04-19 at 3 32 56 PM" src="https://user-images.githubusercontent.com/101145325/233169414-306b6508-e48f-4d33-a5f9-05aad93f0a7a.png">
<img width="578" alt="Screenshot 2023-04-19 at 3 33 00 PM" src="https://user-images.githubusercontent.com/101145325/233169430-2a036850-401d-4dbf-82d4-a61fd376c604.png">

Now, without `{ rank: false }`:
<img width="570" alt="Screenshot 2023-04-19 at 3 37 55 PM" src="https://user-images.githubusercontent.com/101145325/233169605-ac6feca0-0866-41d9-bb18-61ca2fd63943.png">
<img width="850" alt="Screenshot 2023-04-19 at 3 38 04 PM" src="https://user-images.githubusercontent.com/101145325/233169650-3f4ce8da-6d98-4013-b9ad-fc820ba26c94.png">

The number of active handles increases a lot over time, and the memory used also increases, but more slowly. I think this explains why we are seeing our server crash in production

Any idea what could be happening? After we last talked I made some changes and we are basically having one single client being used for each chain. We build each client once, and then re-use it

**Reply by 0xsambugs** [reply]
Ok, I have some more data. Even with `{ rank: false }`, we are still leaking memory. Here, you can see how we started leaking memory when we started using viem vs ethers. It was slower than with `rank` activated, but still there. Just as an FYI, the only change was that we started using viem for multicalls

![Screenshot 2023-04-20 at 6 46 55 AM](https://user-images.githubusercontent.com/101145325/233329011-20aca7f1-899d-43c8-8a90-8022308ccf4b.png)

**Reply by jxom** [reply]
Interesting. Would you be able to spin up a minimal reproduction in a separate repository (ideally with no other dependencies). I can take a look into it, and run some memory tests.

**Reply by 0xsambugs** [reply]
Will do my best 🫡

**Reply by 0xsambugs** [reply]
Hey @jxom , sorry for the delay but it was very hard to figure out where the problem was. Just wanted to confirm that the issue wasn't with viem, it was with Node JS. With the use of viem, we started using Node 18's native fetch, but apparently they [introduced a memory leak in 18.13 ](https://github.com/nodejs/node/issues/46435)
[It was fixed in 18.15.0](https://github.com/nuxt/nuxt/issues/12384#issuecomment-1506740463), but [re-introduced in 18.16.0](https://github.com/nuxt/nuxt/issues/12384#issuecomment-1506764535). It was very hard to figure all this out 😅 
The fix was just hardcoding Node's version to 18.15

And going back to the reason why I created the issue. Did you get a change to add a chainId check to custom transports so that they work with fallback transports?

**Comment by ppastor06** [comment]
very interesting

**Comment by just2102** [comment]
basically there isn't any point in having a fallback() mechanism since it doesn't work when switching/adding chains.
so if an RPC I use is down my app will work fine ONLY if users have my chain in their wallet, which is unreliable.
